catroid/src/main/java/org/catrobat/catroid/ui/cloudmessaging/CustomFirebaseMessagingService.java

```diff
@@ -0,0 +1,115 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.ui.cloudmessaging;
+
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.media.RingtoneManager;
+import android.net.Uri;
+import android.support.v4.app.NotificationCompat;
+import android.util.Log;
+
+import com.google.firebase.messaging.FirebaseMessagingService;
+import com.google.firebase.messaging.RemoteMessage;
+
+import org.catrobat.catroid.R;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Map;
+
+public class CustomFirebaseMessagingService extends FirebaseMessagingService {
+
+	public static final String TAG = CustomFirebaseMessagingService.class.getSimpleName();
+
+	@Override
+	public void onMessageReceived(RemoteMessage remoteMessage) {
+		RemoteMessage.Notification notification = remoteMessage.getNotification();
+
+		String title = notification.getTitle();
+		String message = notification.getBody();
+
+		Map<String, String> customData = remoteMessage.getData();
+
+		if (!customData.containsKey("image") || !customData.containsKey("link") || isStringNullOrEmpty(title) || isStringNullOrEmpty(message)) {
+			Log.e(TAG, "Cloud Notification Failed due to missing parameter(s)");
+			return;
+		}
+
+		String image = customData.get("image");
+		String link = customData.get("link");
+
+		Bitmap imageBitmap = getBitmap(image);
+
+		sendNotification(title, message, imageBitmap, link);
+	}
+
+	private void sendNotification(String title, String message, Bitmap image, String link) {
+		Log.d(TAG, "Title : " + title);
+		Log.d(TAG, "Body : " + message);
+		Log.d(TAG, "link : " + link);
+		Log.e(TAG, "image : " + image);
+
+		Uri webPage = Uri.parse(link);
+		Intent intent = new Intent(Intent.ACTION_VIEW, webPage);
+		PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, 0);
+		Uri defaultSoundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
+
+		NotificationCompat.Builder notification = new NotificationCompat.Builder(this)
+				.setSmallIcon(R.drawable.ic_launcher)
+				.setLargeIcon(image)
+				.setContentTitle(title)
+				.setContentText(message)
+				.setAutoCancel(true)
+				.setSound(defaultSoundUri)
+				.setContentIntent(pendingIntent);
+
+		NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+		notificationManager.notify(1, notification.build());
```

Comment:
id 1 will always overwrite the current notification.

There is a [StatusBarNotificationManager.java](https://github.com/Catrobat/Catroid/blob/be9b08440d257bdb9adbd6cf142aae5b41a8e8d0/catroid/src/main/java/org/catrobat/catroid/utils/StatusBarNotificationManager.java) but I'm not sure if it's well suited. It handles any kind of Notification and I'm definitely not happy with that. The id is also not thread safe and (maybe) should be saved persistently.

We will have to improve this and the StatusBarNotificationManager to work well with notifications. However, for now just keep this code and start the notification ids at 1000. No thread safety or persistence needed.